### ISSUE TITLE
Incorrect type inference when reassigning a parameter using pathlib.Path

### ISSUE THREAD

**Issue by willfrey** [microsoft/pylance-release#348]
## Environment data

-   Language Server version: v2020.9.0
-   OS and version: macOS Catalina v10.15.6
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.5

## Expected behaviour

Pylance should recognize that reassigning `path = pathlib.Path(path)` promotes `path` to be an instance of `pathlib.Path`, which is stronger than `_PathLike[str]`.

This example is contrived but demonstrates the point.

```py3
import os
import pathlib
from typing import Union, TYPE_CHECKING


def read_text(path: Union[str, "os.PathLike[str]"]) -> str:
    path = pathlib.Path(path)

    if TYPE_CHECKING:
        # Pylance should agree with mypy.
        # mypy says: Revealed type is 'pathlib.Path*'
        reveal_type(path)

    # Pylance shouldn't complain about `read_text()` being an unknown member.
    return path.read_text()
```

## Actual behaviour

Pylance correctly infers that `path` can no longer be an instance of `str` but fails to recognize that `path` is an instance of `pathlib.Path`.

```py3
import os
import pathlib
from typing import Union, TYPE_CHECKING


def read_text(path: Union[str, "os.PathLike[str]"]) -> str:
    path = pathlib.Path(path)

    if TYPE_CHECKING:
        # Pylance says: Type of "path" is "_PathLike[str]"
        # mypy says: Revealed type is 'pathlib.Path*'
        reveal_type(path)

    # Pylance complains: Cannot access member "read_text" for type "_PathLike[str]"
    return path.read_text()
```

## Logs

I'm not sure if logs are helpful here. I can include them if you want them but I'd need to tidy up some of the output from elsewhere in the project I'm working on.

## Code Snippet / Additional information

I recognize that this is easy to work around. I can just pop an assert in there to force the correct type inference, as follows:

```py3
import os
import pathlib
from typing import Union, TYPE_CHECKING


def read_text(path: Union[str, "os.PathLike[str]"]) -> str:
    path = pathlib.Path(path)

    assert isinstance(path, pathlib.Path)

    if TYPE_CHECKING:
        # Pylance says: Type of "path" is "_PathLike[str]"
        # mypy says: Revealed type is 'pathlib.Path*'
        reveal_type(path)

    # Pylance complains: Cannot access member "read_text" for type "_PathLike[str]"
    return path.read_text()
```

It'd be nice if Pylance could figure this out on its own, though!

Thank you for all your work on Pylance!

**Comment by erictraut** [comment]
Thanks for reporting this. I've introduced a fix, and it will be in the next release.

If you're interested in the details... This wasn't related to type inference. It was related to type narrowing for assignments. In this particular case, the type evaluator was correctly determining that the type of the expression `pathlib.Path(path)` was `Path`. It then assigned it to the symbol `path`, which has a declared type of `Union[str, os.PathLike[str]]`. When assigning to a symbol with a declared union type, Pyright's narrowing logic attempted to narrow the union. In this case, it eliminated the `str` from the union, which left `os.PathLike[str]`. However, it could have (and should have) further narrowed this type because `Path` is a narrower type than `os.PathLike[str]`.

That's probably more detail than you wanted, but I figured maybe you'd find it interesting.
